### ISSUE TITLE
[BUG] Variables, once clamped, are not logged

### ISSUE THREAD

**Issue by ms609** [revbayes/revbayes#598]
**Describe the bug**
Clamping a variable causes it to be omitted from log files.

**To Reproduce**
```rb
seed(0)
chars <- readDiscreteCharacterData("simmat/k5obsno5.nex")
taxa <- chars.names()
outgroup <- clade(taxa[1])
nTaxa <- chars.size()
nEdge <- 2 * nTaxa - 3

moves = VectorMoves()
topology ~ dnUniformTopology(taxa, outgroup = outgroup)
moves.append( mvNNI(topology) )
moves.append( mvSPR(topology) )
tree_length ~ dnGamma(shape = 2, rate = 2 / 3)
moves.append( mvScale(tree_length) )

rel_br_lengths ~ dnDirichlet( rep(1.0, nEdge) )
moves.append( mvBetaSimplex(rel_br_lengths) )
moves.append( mvDirichletSimplex(rel_br_lengths) )
br_lengths := rel_br_lengths * tree_length

phylogeny := treeAssembly(topology, br_lengths)

u ~ dnGamma( shape = 1, rate = 1 )
rates := [ [0, u, 1, 1],
           [u, 0, u, 1],
           [1, u, 0, u],
           [1, 1, u, 0]]

Q4prime := fnFreeK(rates)

m_morph[1] ~ dnPhyloCTMC(
    tree = phylogeny,
    Q = Q4prime,
    type = "Standard"
)
m_morph[1].clamp(chars)
m_morph[1].lnProbability()

monitors = VectorMonitors()
monitors.append( mnScreen(printgen = 1, u, tree_length) ) # u is printed to screen
monitors.append( mnFile(filename = "mnFile.log", u, tree_length, printgen = 1) ) # u is printed
monitors.append( mnModel(filename = "mnModel.log", u, printgen = 1,
                         exclude = ["br_lengths", "rel_br_lengths"]) ) # u does not appear

u.clamp(1) # Remove this line and u is logged
mcmc(model(phylogeny), monitors, moves).run(generations = 3)


# Also occurs in mcmcmc:
#mymc3 = mcmcmc(model(phylogeny), monitors, moves, nruns = 2, nchains = 6,
#               moveschedule = "single", swapMethod = "both", 
#               swapMode = "multiple", combine = "mixed").run(generations = 3)
```

In the mcmcmc instance, we see a bonus error:
```
Error:       Executing NNI(topology): PDFs not up-to-date before proposal!
       m_morph[1]: -31531.057172 != -inf    diff = inf
```

**Expected behavior**
_u_ should be recognized as part of the model and logged without being explicitly specified.  Clamping _u_ should not affect whether the variable is logged, as the value is updated during MCMC whether or not the variable is clamped.

If variables under move proposals are not meant to be clamped (?) then this should fail with an error instead.

**Computer info**
Win10, RB devel

**Additional context**
Possibly related: #566

**Comment by bjoelle** [comment]
> the value is updated during MCMC whether or not the variable is clamped

Can you explain this a bit further ? My understanding is that clamping is precisely designed for variables whose value is known and thus should **not** change during the run

**Comment by ms609** [comment]
I wondered whether that might be the case.  I've used `u.clamp()` for two things:
1. to validate the behaviour of the model. For example:
```rb
u.clamp(1)
print(m_morph[1].lnProbability())
u.clamp(100)
print(m_morph[1].lnProbability())
```

2. to ensure that an MCMC run is initialized from a particular value (as in the example above)

Currently, the value of u _does_ change during the run, as a consequence of it being subject to `mvScale`.  If clamped variables should not change, then it would be useful for RevBayes to raise an exception in the case that a user attempts to apply a move operator to a clamped variable (or vice versa).

If `clamp` is unsuitable for the purposes above, is there an alternative way of assigning a particular value to a deterministic variable?

**Comment by bjoelle** [comment]
the function `setValue` allows the value of a node to be set so it can be used for initial values

I'm not sure whether anything actually prevents a clamped variable from changing, I was under the impression that this would give a likelihood of `-Inf` ? using the script above I do not see u moving

**Comment by ms609** [comment]
Thanks, `setValue` is the function I needed. 

And whoops – I must have optimised out the `moves.append(mvScale(u))` line when trying to attain an MRE!  I'll put it back now.

**Comment by bjoelle** [comment]
ok, so with the move it does change the value of `u` - that doesn't seem right, it should prevent you from doing that I think

**Comment by ms609** [comment]
So, I'd suggest that we:

- [ ] #599 in the documentation
- [ ] Throw an exception when a clamped stochastic node attempts to redraw its value

Are there other functions that might attempt to modify the value of a clamped node when it should be fixed?  As `.setValue()` is equivalent to `.unclamp(); .clamp()`, I think `x.clamp(1); x.setValue(2)` is legitimate.

Possibly desirable:
- [ ] Warn the user when a move is defined on a clamped node.  
- [ ] Warn the user when a node that is subject to a move operator is clamped

We should probably not throw an exception in the latter cases – and perhaps not even a warning – as there is nothing intrinsically wrong with

```rb
x.clamp(1)
someVector.append(mvScale(x)) # We don't know whether this will ever be applied to x
x.unclamp()
# x not clamped; moves are legitimate
mcmc(model(x), moves = someOtherVector).run()
```

... even if it is difficult to imagine a scenario in which this situation might arise.

**Comment by bjoelle** [comment]
in your last case I would say that `model` is the one that needs to throw the exception here, but if that's not possible I would rather be conservative and just stop moves from applying to clamped nodes (or add a prompt that requires the user to actually say "yes, I know what I'm doing")
my experience is that warnings are often unseen/ignored by users especially if they happen in the middle of a bunch of other text like loading data messages, initialization messages and whatever else

**Comment by ms609** [comment]
I agree that an exception at this point would be the most robust approach (though presumably the responsible element would be `mcmc` rather than `model`, which does not know the content of `someOtherVector`).

**Comment by davidcerny** [comment]
I'm marking this as partially fixed by #600. The problem is that the script from @ms609's opening post, in which we place a move on `u` _before_ clamping it, still runs without issues, causing `u` to be changed and monitored just like any other stochastic node. The desired outcome – throwing `   Error:	Cannot add the clamped node 'u' to proposal Scaling()` – only occurs if we place a move on `u` _after_ clamping it. This is unfortunate: ideally, we'd want `model` to know which nodes can or cannot have moves placed on them, and the order in which we specify things in the script should not matter.
